verifier/src/tls_doc.rs

```diff
@@ -0,0 +1,168 @@
+    signed_tls: SignedTLS,
+    committed_tls: CommittedTLS,
+}
+
+impl TLSDoc {
+    pub fn new(signed_tls: SignedTLS, committed_tls: CommittedTLS) -> Self {
+        Self {
+            signed_tls,
+            committed_tls,
+        }
+    }
+
+    /// Verifies the TLS document against the DNS name `dns_name`:
+    /// - end entity certificate was issued to `dns_name` and was valid at the time of the
+    ///   notarization
+    /// - certificate chain was signed by a trusted certificate authority
+    /// - key exchange parameters were signed by the end entity certificate
+    /// - commitment to misc TLS data is correct
+    ///
+    pub fn verify(&self, dns_name: String) -> Result<(), Error> {
+        // Verify TLS certificate chain against local root certs. Some certs in the chain may
+        // have expired at the time of this verification. We verify their validity at the time
+        // of notarization.
+        webpki_utils::verify_cert_chain(
+            &self.committed_tls.tls_cert_chain,
+            self.signed_tls.time(),
+        )?;
+
+        let ee_cert = webpki_utils::extract_end_entity_cert(&self.committed_tls.tls_cert_chain)?;
+
+        self.verify_tls_commitment(&self.committed_tls, self.signed_tls.commitment_to_tls())?;
+
+        //check that TLS key exchange parameters were signed by the end-entity cert
+        webpki_utils::verify_sig_ke_params(
+            &ee_cert,
+            &self.committed_tls.sig_ke_params,
+            self.signed_tls.ephemeral_ec_pubkey(),
+            &self.committed_tls.client_random,
+            &self.committed_tls.server_random,
+        )?;
+
+        webpki_utils::check_dns_name_present_in_cert(&ee_cert, dns_name)?;
+
+        Ok(())
+    }
+
+    /// Verifies the commitment to misc TLS data
+    fn verify_tls_commitment(
+        &self,
+        committed_tls: &CommittedTLS,
+        commitment: &[u8; 32],
+    ) -> Result<(), Error> {
+        if blake3(&committed_tls.serialize()?) != *commitment {
+            return Err(Error::CommittedTLSCheckFailed);
+        }
+        Ok(())
+    }
+
+    pub fn signed_tls(&self) -> &SignedTLS {
+        &self.signed_tls
+    }
+
+    pub fn committed_tls(&self) -> &CommittedTLS {
+        &self.committed_tls
+    }
+}
+
+/// an x509 certificate in DER format
+pub type CertDER = Vec<u8>;
+
+/// Misc TLS data which the User committed to before the User and the Notary engaged in 2PC
+/// to compute the TLS session keys
+///
+/// The User should not reveal `tls_cert_chain` because the Notary would learn the webserver name
+/// from it. The User also should not reveal `signature_over_ephemeral_key` to the Notary, because
+/// for ECDSA sigs it is possible to derive the pubkey from the sig and then use that pubkey to find out
+/// the identity of the webserver.
+//
+/// Note that there is no need to commit to the ephemeral key because it will be signed explicitely
+/// by the Notary
+#[derive(Serialize, Clone)]
+pub struct CommittedTLS {
```

Comment:
makes sense about HandshakeData and HandshakeCommitment.
Are you sugesting that what is now `SignedTLS` should be called `SignedHandshakeCommitment`? That wouldn't be accurate since `SignedTLS` contains more than a commitment.